cpp/oneapi/dal/array.hpp

```diff
@@ -229,98 +262,120 @@ class array {
     ///
-    /// @tparam ConstDeleter The type of a deleter used to free.
-    ///                      The deleter implements `void operator()(const T*)`` member function.
+    /// @param data         The shared pointer to externally-allocated memory block.
+    /// @param count        The number of elements of type :literal:`Data` in the memory block.
+    explicit array(const std::shared_ptr<const T>& data, std::int64_t count)
+            : impl_(new impl_t(detail::default_host_policy{}, data, count)) {
+        update_data(data.get(), count);
+    }
+
+#ifdef ONEDAL_DATA_PARALLEL
+    /// Creates a new array instance that shares ownership with the user-provided shared pointer.
     ///
-    /// @param data          The immutable memory block pointer to be assigned inside the array
-    /// @param count         The number of elements of type :literal:`T` into the block
-    /// @param deleter       The object used to free :literal:`Data`.
-    /// @param dependencies  Events indicating availability of the :literal:`Data` for reading or writing.
-    template <typename ConstDeleter>
+    /// @param queue        The SYCL* queue object.
+    /// @param data         The shared pointer to externally-allocated memory block.
+    /// @param count        The number of elements of type :literal:`Data` in the memory block.
+    /// @param dependencies Events that indicate when :literal:`Data` becomes ready to be read or
+    ///                     written
     explicit array(const sycl::queue& queue,
-                   const T* data,
+                   const std::shared_ptr<const T>& data,
                    std::int64_t count,
-                   ConstDeleter&& deleter,
                    const sycl::vector_class<sycl::event>& dependencies = {})
-            : impl_(new impl_t(data, count, std::forward<ConstDeleter>(deleter))) {
-        update_data(impl_.get());
+            : impl_(new impl_t(detail::data_parallel_policy{ queue }, data, count)) {
+        ONEDAL_ASSERT(sycl::get_pointer_type(data.get(), queue.get_context()) !=
+                      sycl::usm::alloc::unknown);
+        update_data(data.get(), count);
         sycl::event::wait_and_throw(dependencies);
     }
 #endif
 
     /// An aliasing constructor: creates a new array instance that stores :literal:`Data` pointer,
-    /// assigns the pointer to the ownership structure of :literal:`ref` to the new instance.
-    /// Array returns :literal:`Data` pointer as its mutable or immutable block depending
-    /// on the :literal:`Data` type.
+    /// assigns the pointer to the ownership structure of :literal:`ref` to the new instance. Array
+    /// returns :literal:`Data` pointer as its mutable or immutable block depending on the
+    /// :literal:`Data` type.
     ///
-    /// @tparam ref    The type of elements in the referenced array.
-    /// @tparam data    Either :literal:`T` or $const T$ type.
+    /// @tparam Y    The type of elements in the referenced array.
```

Comment:
The problem is that `Y` is already in the code, but not reflected in doxygen. The old doxygen comment that was fixed by this change is just incorrect and not displayed in docs

Reply:
> Every doxygen change I think shall trigger spec update since they must be synchronized.

This creates overhead for each PR, I'd propose to have periodical synchronization:
- We need to keep in mind `dalapi` tool, the changes in tool need to be propagated to spec first
- After that we'll copy paste all headers to spec removing implementation, ideally this should be automated

Reply:
Ok